### PR TITLE
chore: trigger run notifications

### PR DESCRIPTION
**Pull request by astone123**
- Closes https://github.com/cypress-io/cypress/issues/26788

### Additional details

This PR adds a new class of actions that allows us to trigger different kinds of notifications. These actions are used in `RelevantRunsDataSource` to send notifications to the user when it is appropriate based on the requirements in the ClickUp. See the related issue for links to the ClickUp and Figma.

I'm using the `chore` prefix in the PR title just so that the changelog step doesn't yell at me. These messages will all be squashed into one commit when we merge the feature branch into develop.

### Steps to test

1. Run Cypress with a project that is connected to Cloud and has at least one run recorded
2. Click "Enable desktop notifications" in the banner at the top of the page (if you've dismissed this banner before you'll need to set `desktopNotificationsEnabled: null` in your global `state.json` file)
3. Break a test in the project (so that the run will fail)
4. Kick off a new test run for the project and leave the Cypress app running
5. Verify that after the first test fails that you see a push notification saying that your run started failing
6. Verify that clicking the notification takes you to the debug page and focuses the run in question
7. Verify that after the run completes you see another push notification saying that your run has failed

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: packages/app/cypress/e2e/settings.cy.ts [hunk 1]
```diff
@@ -459,7 +459,7 @@ describe('App: Settings', () => {
       cy.wait(200)
 
       cy.withCtx((ctx) => {
-        expect(setPreferencesStub).to.have.been.calledWith(JSON.stringify({ notifyWhenRunCompletes: ['passed', 'canceled'] }), 'global')
+        expect(setPreferencesStub).to.have.been.calledWith(JSON.stringify({ notifyWhenRunCompletes: ['passed', 'cancelled'] }), 'global')
```
Comment (warrensplayer) [inline packages/app/cypress/e2e/settings.cy.ts:476]: I looked this up and I found that `cancelled` with two `l`s is the British preferred spelling.  Why did you change this?

File: packages/data-context/src/actions/ElectronActions.ts [hunk 2]
```diff
@@ -4,6 +4,13 @@ import type { DataContext } from '..'
 import _ from 'lodash'
 import path from 'path'
 import assert from 'assert'
+import debugLib from 'debug'
+
+const debug = debugLib('cypress:data-context:ElectronActions')
+
+// Declare notification variable globally so that it doesn't get garbage collected
+// before the user clicks on the notification
+let notification
```
Comment (warrensplayer) [inline packages/data-context/src/actions/ElectronActions.ts]: If more than one notification is triggered, will the older notifications get garbage collected after a while since they are no longer being referenced?  I am not sure how Javascript GC works, but using the `let` is probably less aggressive than reclaiming variables declared in a function scope.  Can you test out sending multiple notifications, waiting for a bit, then trying to click a notification that is not the most recent one?

If needed, one solution I was just thinking of is storing them in an array and then removing them from the array on the `click` handler and on a `close` handler.  It could also clear them after some timeout as well if needed. But don't go to that extra trouble if not needed.

File: packages/app/src/settings/device/NotificationSettings.vue [hunk 3]
```diff
@@ -127,7 +127,7 @@ const listRef = ref(props.gql.localSettings.preferences.notifyWhenRunCompletes)
 const statuses = [
   { id: 'passed', label: t('settingsPage.notifications.passed') },
   { id: 'failed', label: t('settingsPage.notifications.failed') },
-  { id: 'canceled', label: t('settingsPage.notifications.canceled') },
+  { id: 'cancelled', label: t('settingsPage.notifications.canceled') },
```
Comment (warrensplayer) [inline packages/app/src/settings/device/NotificationSettings.vue:165]: Same thing here with spelling. Also, if you keep the two `l`s, the actual i18n key only has one `l`.  I still think we should just use one `l`.

Comment (jordanpowell88) [inline packages/app/src/settings/device/NotificationSettings.vue:165]: we should change the i18n key to `cancelled` here too

File: packages/data-context/src/actions/NotificationActions.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        // failures, then it just started failing.
+        if ((cachedRun.runNumber === newRun.runNumber) && (cachedRun.totalFailed === 0 && newRun.totalFailed > 0)) {
+          this.sendRunFailingNotification(newRun.runNumber)
+        } else {
+          this.sendRunStartedNotification(newRun.runNumber)
+        }
+
+        break
+      default:
+        if (newRun.status) {
```
Comment (jordanpowell88) [inline packages/data-context/src/actions/NotificationActions.ts]: why do we need this `if` here? Won't it never get here in the first place without a `newRun.status` since thats the switch case?

File: packages/data-context/src/sources/RelevantRunsDataSource.ts [hunk 5]
```diff
@@ -243,11 +245,18 @@ export class RelevantRunsDataSource {
...
 
+      debug('current cache: %o, new values: %o', this.#cached, toCache)
+
       //TODO is the right thing to invalidate?  Can we just invalidate the runsByCommitShas field?
       const projectSlug = await this.ctx.project.projectId()
 
       await this.ctx.cloud.invalidate('Query', 'cloudProjectBySlug', { slug: projectSlug })
 
+      // If the cache is empty, then we're just starting up. Don't send notifications
+      if (this.#cached.all.length !== 0 && this.#cached.all[0] && toCache.all[0] && !isEqual(toCache.all[0], this.#cached.all[0])) {
```
Comment (warrensplayer) [inline packages/data-context/src/sources/RelevantRunsDataSource.ts]: Do we want to only focus on the first run, or go through the cache and call the `maybeSendRunNotification`  for each pair of runs (cached vs new) that are not equal?  

An example scenario would be when multiple runs are kicked off for one commit.

File: packages/types/src/preferences.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import type { BannersState, Editor, MajorVersionWelcomeDismissed } from '.'
 
-type NotifyWhenRunCompletes = 'passed' | 'failed' | 'canceled' | 'errored'
+export type NotifyWhenRunCompletes = 'passed' | 'failed' | 'cancelled' | 'errored'
```
Comment (warrensplayer) [inline packages/types/src/preferences.ts:3]: Same comment here about spelling of `canceled` vs `canclled`

File: packages/data-context/src/actions/NotificationActions.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  sendRunFailingNotification (runNumber: number): void {
+    if (this.notifyWhenRunStartsFailingPreference !== true) {
+      debug('notifyWhenRunStartsFailing not true, skipping notification for run #%s', runNumber)
+
+      return
+    }
+
+    this.ctx.actions.electron.showSystemNotification(this.projectTitle, `Run #${runNumber} has started failing`, () => this.onNotificationClick(runNumber))
```
Comment (jordanpowell88) [inline packages/data-context/src/actions/NotificationActions.ts]: I think we should create a `sendNotification` method since all 3 of these methods are calling the same `showNotification` action. Something like this that each of these methods would call:

```ts
sendNotification(runNumber: number) {
   this.ctx.actions.electron.showSystemNotification(this.projectTitle, `Run #${runNumber} has started failing`, () => this.onNotificationClick(runNumber))
}
```

File: packages/data-context/src/sources/RelevantRunsDataSource.ts (hunk 5 above)
Comment (warrensplayer) [reply packages/data-context/src/sources/RelevantRunsDataSource.ts]: Also, I think because of the check for `this.#cached.all.length !== 0` that is there to prevent it from firing when the app first starts up, then i think this would not fire if the app had been open a while with no runs being fetched, and then a new run comes in.  This scenario is one I don't think needs too much time on and could be filed as a fast-follow in my opinion.  The users that will get the most value from this will be ones that already have runs.

Comment (jordanpowell88) [inline packages/data-context/src/sources/RelevantRunsDataSource.ts]: I'm not seeing a test for this case. It's a bit hard to read so want to make sure we have good test coverage around this logic.

**Review by jordanpowell88 — CHANGES_REQUESTED** [review]
Overall excellent job with this. Just a few small suggestions

File: packages/data-context/src/actions/ElectronActions.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/ElectronActions.ts]: I had the same thought, but using a `Map` instead of an array. We could give each one a uuid and do `map.set(uuid, notification)`. When the click handler is called, we could just do `map.delete(uuid)`.

I don't think we ever want to clear them as long as the App is open, since at least in MacOS, you can stack notifications forever:

image

Perhaps someone comes back after the weekend and wants to see what failed when they pushed code and clocked off on Friday afternoon.

File: packages/data-context/src/actions/NotificationActions.ts (hunk 4 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/NotificationActions.ts]: I'm guessing `switch('PASSED')`, or anything other than `RUNNING` would mean we fall through to this case.

It looks like we could simply do

```js
if (!newRun.status) {
  return
}
```

before the switch statement.

File: packages/data-context/src/sources/RelevantRunsDataSource.ts [hunk 8]
```diff
@@ -243,11 +245,18 @@ export class RelevantRunsDataSource {
...
       debug('Values changed')
 
+      debug('current cache: %o, new values: %o', this.#cached, toCache)
+
       //TODO is the right thing to invalidate?  Can we just invalidate the runsByCommitShas field?
       const projectSlug = await this.ctx.project.projectId()
 
       await this.ctx.cloud.invalidate('Query', 'cloudProjectBySlug', { slug: projectSlug })
 
+      // If the cache is empty, then we're just starting up. Don't send notifications
```
Comment (lmiller1990) [inline packages/data-context/src/sources/RelevantRunsDataSource.ts:255]: I found this comment a little confusing - it says "if the cache is empty" but the condition is actually the opposite, is the cache NOT empty.

I wonder if we could benefit from a variable here?

```js
const cacheContainsRun = (this.#cached.all.length !== 0 && this.#cached.all[0] && toCache.all[0] && !isEqual(toCache.all[0], this.#cached.all[0])) 

if (cacheContainsRun) {
  ...
}
```

Also can't it just be

```
const cacheContainsRun = (this.#cached.all.length > 0 && !isEqual(toCache.all[0], this.#cached.all[0])) 
```

?

**Review by lmiller1990 — COMMENTED** [review]
Some high level questions, I'll give this a test now.

**Comment by lmiller1990** [comment]
I tried it out. I only got one notification, which I suspect is correct, considering my project is tiny - by the time the App polled the Cloud, the run had failed AND the run had completed.

Here is how it looks on my machine (Ubuntu).


https://github.com/cypress-io/cypress/assets/19196536/a7cc6ee8-552d-4b34-a3ba-ab4bee24a735

File: packages/app/cypress/e2e/settings.cy.ts (hunk 1 above)
Comment (astone123) [reply packages/app/cypress/e2e/settings.cy.ts:476]: I changed this to conform with the `CloudRunStatus` enum [here](https://github.com/cypress-io/cypress/blob/4c823909fe45fe29b7585b54880060ca6a6efb96/packages/graphql/schemas/schema.graphql#L689). We're passing that status directly into the notification functions so we need the keys for the preferences to match the cloud run status enum.

File: packages/app/src/settings/device/NotificationSettings.vue (hunk 3 above)
Comment (astone123) [reply packages/app/src/settings/device/NotificationSettings.vue:165]: I think we should use one `l` too, but if we do that we either need to normalize the `CANCELLED` cloud run status when it comes in, or update the cloud schema to also only use one `l`

File: packages/data-context/src/actions/NotificationActions.ts (hunk 4 above)
Comment (astone123) [reply packages/data-context/src/actions/NotificationActions.ts]: @jordanpowell88 if `newRun.status` is undefined we will still reach the default case. Yes, we could return early if there is no status

File: packages/data-context/src/sources/RelevantRunsDataSource.ts (hunk 8 above)
Comment (astone123) [reply packages/data-context/src/sources/RelevantRunsDataSource.ts:255]: The comment is referring to this condition
```js
this.#cached.all.length !== 0
```

As in, if the current cache is empty, then we don't want to send a notification. If we sent one in this case, then users would get notified about whichever run is the latest from the cloud, even if it is like weeks old. 

The comment could use some clarification. Also up for moving these conditions into a variable. I added these
```js
this.#cached.all[0] && toCache.all[0]
```
to satisfy Typescript, because otherwise those could be undefined. If they are, then we don't want to call `maybeSendRunNotification` anyway.

**Comment by astone123** [comment]
@lmiller1990 that sounds correct... not related to this PR, but the notification should have the Cypress logo in it 🤔

File: packages/data-context/src/actions/ElectronActions.ts (hunk 2 above)
Comment (warrensplayer) [reply packages/data-context/src/actions/ElectronActions.ts]: The map is much better!

File: packages/data-context/src/actions/NotificationActions.ts (hunk 7 above)
Comment (warrensplayer) [reply packages/data-context/src/actions/NotificationActions.ts]: Good idea. It would need to take the message as well since it differs per type of notification.

File: packages/app/src/settings/device/NotificationSettings.vue (hunk 3 above)
Comment (lmiller1990) [reply packages/app/src/settings/device/NotificationSettings.vue:165]: If the cloud is already standardized on two `l`s, I think we should just stick with that. Changing a cloud status sounds like we'd need a database migration, which is quite involved. WDYT @warrensplayer?

**Comment by lmiller1990** [comment]
I will pick this up.

File: packages/data-context/src/actions/ElectronActions.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/ElectronActions.ts]: Actually, we can do better with `Set`. Those are unique by reference. We can just do `set.add(notification)`, and when you click it, `set.delete(notification)`.

File: packages/data-context/src/sources/RelevantRunsDataSource.ts (hunk 8 above)
Comment (lmiller1990) [reply packages/data-context/src/sources/RelevantRunsDataSource.ts:255]: Ok, sure. One last minor one... Given:

```
this.#cached.all.length !== 0 && this.#cached.all[0]
```

I think `this.#cached.all.length !== 0` is not needed, since `this.#cached.all[0]` implies that the length is not 0.

File: packages/types/src/preferences.ts (hunk 6 above)
Comment (lmiller1990) [reply packages/types/src/preferences.ts:3]: We may need to opt for `cancelled`, since this is what the Cloud is sending and it's not at all practical to change their code at this point. We've got a mix of both now, which isn't really ideal.

File: packages/data-context/src/actions/ElectronActions.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/ElectronActions.ts]: https://github.com/cypress-io/cypress/pull/26974/commits/58ee97d152e18c119a6821e32d94e2547ae1836c

File: packages/data-context/src/sources/RelevantRunsDataSource.ts (hunk 5 above)
Comment (lmiller1990) [reply packages/data-context/src/sources/RelevantRunsDataSource.ts]: https://github.com/cypress-io/cypress/pull/26974/commits/95270660c9030c1a8b5677dbf22d16a2985ffa6e

`this.#cached.all[0]` implies `length !== 0`, so I removed that part.

File: packages/types/src/preferences.ts (hunk 6 above)
Comment (lmiller1990) [reply packages/types/src/preferences.ts:3]: I think we should get feature completed and we can figure out the `canceled` vs `cancelled` stuff at the end, it's definitely confusing, we need to pick one and stick with it, I'd recommend whatever has the least risk, which is likely conforming with the Cloud's definition (double `l`).

File: packages/data-context/src/actions/NotificationActions.ts (hunk 4 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/NotificationActions.ts]: https://github.com/cypress-io/cypress/pull/26974/commits/dcf4794f5ac4c2afda4f313fa270783b80035a82

File: packages/data-context/src/actions/NotificationActions.ts (hunk 7 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/NotificationActions.ts]: I did it here: https://github.com/cypress-io/cypress/pull/26974/commits/dcf4794f5ac4c2afda4f313fa270783b80035a82

File: packages/data-context/src/sources/RelevantRunsDataSource.ts (hunk 5 above)
Comment (lmiller1990) [reply packages/data-context/src/sources/RelevantRunsDataSource.ts]: I think we do actually, search for `expect(maybeSendRunNotificationStub)` in the tests - it looks like Adam just updated existing tests that already run this function to assert it's doing the correct thing. This seems reasonable to me.

**Comment by lmiller1990** [comment]
End to End testing this feature is challenging. I tried a few different scenarios locally, and it seems to be working as expected. On Ubuntu, I did not see the Cypress icon in the notification - I'm not sure why.

Either way, I think there will be some minor changes when we do the final QA. I'm going to put together some scenarios and test projects we can use to verify it's working correctly. The code as it stands now looks pretty good to me.

**Review by lmiller1990 — APPROVED** [review]
End to End testing this feature is challenging. I tried a few different scenarios locally, and it seems to be working as expected. On Ubuntu, I did not see the Cypress icon in the notification - I'm not sure why.

Either way, I think there will be some minor changes when we do the final QA. I'm going to put together some scenarios and test projects we can use to verify it's working correctly. The code as it stands now looks pretty good to me.

**Review by jordanpowell88 — APPROVED** [review]

**Review by warrensplayer — APPROVED** [review]
LGTM
